README.md
Name
====

lua-nginx-osm - Lua Tirex client driver for the ngx_lua based on the cosocket API

Status
======

Current version is 0.43, 21, October, 2013.
This library is considered active development status.

Description
===========

This Lua library is a Tirex client driver for the ngx_lua nginx module:

http://wiki.nginx.org/HttpLuaModule

This Lua library takes advantage of ngx_lua's cosocket API, which ensures 
100% nonblocking behavior.

It also includes utility to handle metatile, URIs in Lua language.
These utility is not depend on nginx, means pure lua implementation.

Note that at least [ngx_lua 0.8.1](https://github.com/chaoslawful/lua-nginx-module/tags) is required.

If you use Ubuntu Linux 12.04 (LTS) and after, there is a PPA(private package archive) for you.
http://launchpad.net/~osmjapan/+archive/ppa
Please see the above page for detail instructions.

Synopsis
========

    lua_package_path "/path/to/lua-nginx-osm/?.lua;;";
    lua_shared_dict osm_tirex 10m; ## mandatory to use osm.tirex module

    server {
        location /example {
            content_by_lua '
                local tirex = require "osm.tirex"
                local tile = require "osm.tile"
                local data = require "osm.data"
                
                -- --------------------------------------------------
                -- check uri
                -- --------------------------------------------------
                local uri = ngx.var.uri
                local map = "example"
                local x, y, z = tile.get_cordination(uri, map, ".png")
                if not x then
                    return ngx.exit(ngx.HTTP_FORBIDDEN)
                end
                
                -- check x, y, z range
                local max_zoom = 18
                local min_zoom = 5
                if not tile.check_integrity_xyzm(x, y, z, minz, maxz) then
                    return ngx.exit(ngx.HTTP_FORBIDDEN)
                end
                
                -- check x, y, z supported to generate
                local region = data.get_region("japan")
                if not osm_tile.is_inside_region(region, x, y, z)
                    -- try upstream server?
                    return ngx.exit(ngx.HTTP_FORBIDDEN)
                end
                
                -- try renderd file
                local tilefile = tile.xyz_to_metatile_filename(x, y, z)
                local tilepath = tirex_tilepath.."/"..map.."/"..tilefile
                local png, err = tile.get_tile(tilepath, x, y, z)
                if png then
                    ngx.header.content_type = "image/png"
                    ngx.print(png)
                    return ngx.OK
                end
    
                -- ask tirex to render it
                local ok = tirex.send_request(map, x, y, z)
                if not ok then
                    return ngx.exit(ngx.HTTP_INTERNAL_SERVER_ERROR)
                end

                -- get tile image from metatile
                local png, err = tile.get_tile(tilepath, x, y, z)
                if png then
                    ngx.header.content_type = "image/png"
                    ngx.print(png)
                    return ngx.OK
                end
                return ngx.exit(ngx.HTTP_NOT_FOUND)
            ';
        }
    }

Tile Methods
=======

get_cordination
-----------------

**syntax:** *x, y, z = get_cordination(uri, map, ext)*

Retrive x/y/z from uri path.
If client GET uri   /example/9/3/1.png   then example map, z =9, x=3 and y=1
and extension is png.

xyz_to_metatile_filename
-------------------------

**syntax:** *filename = osm.tile.xyz_to_metatile_filename(x, y, z)*

Generate metatile filename from x/y/z cordination.
 
get_tile
--------

**syntax:** *png, err = osm.tile.get_tile(tilepath, x, y, z)*

Get chunk of png image data of x/y/z cordination from metatile tilepath.

check_integrity_xyzm
----------------------

**syntax:** *ok = osm.tile.check_integrity_xyzm(x, y, z, minz, maxz)*

Check whether x/y/z integrity.
Tile x/y/z definition details are in
   https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames

check_integrity_xyz
----------------------

**syntax:** *ok = osm.tile.check_integrity_xyzm(x, y, z)*

Same as check_integrity_xyzm() but don't check z range.

is_inside_region
--------

**syntax:** *include = osm.tile.is_inside_region(region, x, y, z)*

Check x/y/z cordination is located and inside of region.
region should be get from 'osm.data'


Data methods
============

**syntax:** * region = data.get_region("japan"))*
 
Get region definition table of argument country/area.
This can use for is_inside_region() method.

Now provide following area/country data:

    japan
    asia
    world


Tirex Methods
=======

request
-------------

**syntax:** *result = osm.tirex.request(map, x, y, z1, z2, priority)*

Request enqueue command to rendering map 'map' with x/y/z1 cordination and 
priority. And also request to render in background between zoom z1 to z2. 
If request fails return nil.

When z1 == z2, just ask to render single tile.


cancel
-------------

**syntax:** *result = osm.tirex.cancel(map, x, y, z1, z2, priority)*

Request dequeue command to rendering map 'map' with x/y/z1 cordination and 
priority.And also request to cancel in background between zoom z1 to z2. 
If request fails return nil.


ping
-------------

**syntax:** *result = osm.tirex.ping()*

Request ping command. If request fails return nil.


Obsolete functions
=============

send_request
-------------

**syntax:** *result = osm.tirex.send_request(map, x, y, z)*

Request enqueue command to rendering map 'map' with x/y/z cordination.
If request fails return nil.


enqueue_request
-------------

**syntax:** *result = osm.tirex.enqueue_request(map, x, y, z, priority)*

Request enqueue command to rendering map 'map' with x/y/z cordination and 
priority.
If request fails return nil.

dequeue_request
-------------

**syntax:** *result = osm.tirex.dequeue_request(map, x, y, z, priority)*

Request dequeue command to rendering map 'map' with x/y/z cordination and 
priority.
If request fails return nil.


TODO
====

* build more data definitions

* and more on issue tracker.


Community
=========

English Mailing List
--------------------

The [tile-serving](https://lists.openstreetmap.org/lists/tile-serving) mailing list is for English speakers.
It is for all topic about tile serving development of openstreetmap, not only this project.

Web Chat 
--------------------

The [osmfj-devel](http://lingr.com/signup?letmein=osmfj_devel) web chat is in Japanese/English.
It is a chat room mainly for OSM Japan site and related software development.


Bugs and Patches
================

Please report bugs or submit patches by

1. creating a ticket on the [GitHub Issue Tracker](http://github.com/miurahr/lua-nginx-osm/issues),

1. There are known problem that Tirex cannot response properly so we need to patch to tirex.

  https://trac.openstreetmap.org/ticket/4869
  
  If you use Tirex 0.4.1(original)  or  tirex-0.4.1ppa4 and below, you need to patch to tirex.
  Here is a patch file in misc/tirex-peer.diff.

Author
======

Hiroshi Miura <miurahr@osmf.jp>, OpenStreetMap Foundation Japan

Copyright and License
=====================

Hiroshi Miura, 2013 
OpenStreetMap Foundation Japan, 2013

Distributed under GPLv3 License.

See Also
========
* the ngx_lua module: http://wiki.nginx.org/HttpLuaModule




[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/miurahr/lua-nginx-osm/trend.png)](https://bitdeli.com/free "Bitdeli Badge")



debian/changelog
lua-nginx-osm (0.44) precise; urgency=low

  * Fix north-america data missing.
  * Fix missing data in package.

 -- Hiroshi Miura <miurahr@linux.com>  Mon, 14 Apr 2014 08:03:25 +0900

lua-nginx-osm (0.43) precise; urgency=low

  * Fix debian dependency error.

 -- Hiroshi Miura <miurahr@linux.com>  Mon, 30 Sep 2013 20:43:38 +0900

lua-nginx-osm (0.42) precise; urgency=low

  * Update dependency to allow nginx-openresty for nginx/lua server.

 -- Hiroshi Miura <miurahr@linux.com>  Mon, 30 Sep 2013 07:57:03 +0900

lua-nginx-osm (0.41) precise; urgency=low

  * Excape PPA problem with incrementing version.

 -- Hiroshi Miura <miurahr@linux.com>  Sun, 22 Sep 2013 12:47:08 +0900

lua-nginx-osm (0.40) precise; urgency=low

  * New version: add many region data.

 -- Hiroshi Miura <miurahr@linux.com>  Sun, 15 Sep 2013 11:04:55 +0900

lua-nginx-osm (0.30-1) precise; urgency=low

  * New version: background request feature.

 -- Hiroshi Miura <miurahr@linux.com>  Sat, 29 Jun 2013 08:28:51 +0900

lua-nginx-osm (0.20-1) precise; urgency=low

  * New version: fix tirex socket problem.

 -- Hiroshi Miura <miurahr@linux.com>  Sun, 02 Jun 2013 17:35:25 +0900

lua-nginx-osm (0.19-1) precise; urgency=low

  * remove Tirex buf workaround
  * test for 0.20 version

 -- Hiroshi Miura <miurahr@linux.com>  Sun, 02 Jun 2013 14:46:07 +0900

lua-nginx-osm (0.10-2) precise; urgency=low

  * Add tirex dependency

 -- Hiroshi Miura <miurahr@linux.com>  Sun, 12 May 2013 23:34:46 +0900

lua-nginx-osm (0.10-1) precise; urgency=low

  * Initial release.

 -- Hiroshi Miura <miurahr@linux.com>  Sun, 12 May 2013 11:27:52 +0900



osm/tirex.lua
--
-- Lua script for interface Tirex engine
--
--
-- Copyright (C) 2013, Hiroshi Miura
--
--    This program is free software: you can redistribute it and/or modify
--    it under the terms of the GNU Affero General Public License as published by
--    the Free Software Foundation, either version 3 of the License, or
--    any later version.
--
--    This program is distributed in the hope that it will be useful,
--    but WITHOUT ANY WARRANTY; without even the implied warranty of
--    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
--    GNU Affero General Public License for more details.
--
--    You should have received a copy of the GNU Affero General Public License
--    along with this program.  If not, see <http://www.gnu.org/licenses/>.
--

local shmem = ngx.shared.osm_tirex

local udp = ngx.socket.udp
local time = ngx.time
local timerat = ngx.timer.at
local sleep = ngx.sleep

local sub = string.sub
local len = string.len
local find = string.find
local gmatch = string.gmatch
local format = string.format

local pairs = pairs
local unpack = unpack
local tonumber = tonumber
local tostring = tostring
local error = error
local assert = assert
local setmetatable = setmetatable

local osm_tile = require 'osm.tile'

module(...)

_VERSION = '0.30'

local tirexsock = 'unix:/var/run/tirex/master.sock'
local tirex_cmd_max_size = 512

-- ------------------------------------
-- Syncronize thread functions
--
--   thread(1)
--       get_handle(key)
--       do work
--       store work result somewhere
--       send_signal(key)
--       return result
--
--   thread(2)
--       get_handle(key) fails then
--       wait_singal(key, timeout)
--       return result what thread(1) done
--
--   to syncronize amoung nginx threads
--   we use ngx.shared.DICT interface.
--   
--   Here we use ngx.shared.osm_tirex
--   you need to set /etc/conf.d/lua.conf
--      ngx_shared_dict osm_tirex 10m; 

--   status definitions
--    key is not exist: no job exist for its x/y/z
--    key is exist: job exist
--
--       key := <map>:<x>:<y>:<zoom>
--       val := <req> | <result>
--       flag := <status>
--
--       <x>, <y>, <zoom> := <integer>
--       <req> := string: request command string
--       <result> := string: result string
--       <status> := <gothandle> | <request> | <send> | <succeeded> | <failed>
--
--    key will be expired in timeout(sec)
--
-- ------------------------------------
local GOTHANDLE =   0
local REQUEST   = 100
local SEND      = 200
local SUCCEEDED = 300
local FAILED    = 400
local SPECIAL   = 999

-- function: send_signal
-- argument: string key
--           number timeout in sec
--           number flag to send
-- return nil when failed
local function send_signal(key, timeout, flag)
    local ok, err = shmem:set(key, 0, timeout, flag)
    if not ok then
        return nil
    end
    return true 
end

local function round(num, idp)
  return tonumber(format("%." .. (idp or 0) .. "f", num))
end

-- function: wait signal
-- argument: string key
--           number timeout in second
--           number flag to wait
-- return nil if timeout in wait
--
local function wait_signal(key, timeout)
    local timeout = round(timeout, 1) * 10
    for i=0, timeout do
        local val, flag = shmem:get(key)
        if val then
            if flag == SUCCEEDED then
                return true
            elseif flag == FAILED then
                return nil
            else
                -- do nothing
            end
            sleep(0.1)
        else
            return nil
        end
    end
    return nil
end

-- function: serialize_msg
-- argument: table msg
--     hash table {key1=val1, key2=val2,....}
-- return: string
--     should be 'key1=val1\nkey2=val2\n....\n'
--
local function serialize_msg (msg)
    local str = ''
    for k,v in pairs(msg) do
        str = str .. k .. '=' .. tostring(v) .. '\n'
    end
    return str
end

-- function: deserialize_msg
-- arguments: string str: recieved message from tirex
--     should be 'key1=val1\nkey2=val2\n....\n'
-- return: table
--     hash table {key1=val1, key2=val2,....}
local function deserialize_msg (str)
    local msg = {}
    for line in gmatch(str, "[^\n]+") do
        local m,_,k,v = find(line,"([^=]+)=(.+)")
        if  k ~= '' then
            msg[k]=v
        end
    end
    return msg
end

local function get_key(map, mx, my, mz)
    return format("%s:%d:%d:%d", map, mx, my, mz)
end

--
--  if key exist, it returns false
--  else it returns true
--
local function get_handle(key, val, timeout, flag)
    local success,err,forcible = shmem:add(key, val, timeout, flag)
    if success == false then
	if err == 'exists' then
            local prev_val, prev_flag = shmem:get(key)
            local prev_flag = tonumber(prev_flag)
            if prev_flag < SUCCEEDED then
                local prev_msg = deserialize_msg(prev_val)
                local msg = deserialize_msg(val)
                if prev_msg and prev_msg["prio"] > msg["prio"] then
                    shmem:replace(key, val, timeout, flag)
                    return true
		end
	    else
	        return nil
	    end
	else
	    return nil
        end
    else
        return true
    end
    return nil
end

-- function: remove_handle
-- argument: string key
-- return: nil if failed
local function remove_handle(key)
    return shmem:delete(key)
end

-- ========================================================
--  It does not share context and global vals/funcs
--
local tirex_bk_handler
tirex_bk_handler = function (premature)
    local tirexsock = 'unix:/var/run/tirex/master.sock'
    local tirex_cmd_max_size = 512
    local shmem = ngx.shared.osm_tirex
    local REQUEST   = 100
    local SEND      = 200
    local SUCCEEDED = 300
    local FAILED    = 400

    -- here we cannot refer func so define again
    local deserialize_msg = function (str)
        local msg = {}
        for line in gmatch(str, "[^\n]+") do
            local m,_,k,v = find(line,"([^=]+)=(.+)")
            if  k ~= '' then
                msg[k]=v
            end
        end
        return msg
    end

    if premature then
        -- clean up
        shmem:delete('_tirex_handler')
        return
    end

    local udpsock = ngx.socket.udp()
    udpsock:setpeername(tirexsock)
    udpsock:settimeout(0)

    while true do
        -- ngx.select(shmem, udpsock)
        -- send requests first...
        local indexes = shmem:get_keys()
        for key,index in pairs(indexes) do
            local req, flag = shmem:get(index)
            if flag == REQUEST then
                local ok,err=udpsock:send(req)
                if ok then
                    shmem:replace(index, req, 300, SEND)
                end
            end
        end

        sleep(0.1)

        -- then receive response
        udpsock:settimeout(0)
        local data, err = udpsock:receive(tirex_cmd_max_size)
        if data then
            local msg = deserialize_msg(data)
            local index = get_key(msg["map"], msg["x"], msg["y"], msg["z"])
            local res = msg["result"]
            --send_signal to client context
            local ok
            if res == "ok" then
                shmem:set(index, res, 300, SUCCEEDED)
            else
                shmem:set(index, res, 300, FAILED)
            end
        else
            -- err can be 'timeout', 'partial write', 'closed', 
            -- 'buffer too small' or 'out of memory'
            -- do nothing at this time
        end
    end
    udpsock:close()
end

local function background_enqueue_request(map, x, y, z, priority)
    local mx = x - x % 8
    local my = y - y % 8
    local mz = z
    local id = time()
    local priority = tonumber(priority)
    local index = get_key(map, mx, my, mz)
    local req = serialize_msg({
        ["id"]   = tostring(id);
        ["type"] = 'metatile_enqueue_request';
        ["prio"] = priority;
        ["map"]  = map;
        ["x"]    = mx;
        ["y"]    = my;
        ["z"]    = mz})
    local ok = get_handle(index, req, 300, REQUEST)
    if not ok then
        return nil
    end
    local handle = get_handle('_tirex_handler', 0, 0, SPECIAL)
    if handle then
        -- only single light thread can handle Tirex
        timerat(0, tirex_bk_handler)
    end

    return true
end


-- function: send_tirex_request
-- return: resulted msg{}
local function send_tirex_request(req)
    local udpsock = udp()
    udpsock:setpeername(tirexsock)
    local ok,err=udpsock:send(req)
    if not ok then
        udpsock:close()
        return nil
    end
    local data, err = udpsock:receive(tirex_cmd_max_size)
    udpsock:close()
    if not data then
        return nil
    end
    local msg = deserialize_msg(data)
    return msg
end

-- funtion: send_request
-- argument: map, x, y, z
-- return:   true or nil
--
function send_request (map, x, y, z)
    return enqueue_request(map, x, y, z, 1)
end

--[[
Buckets definition in default
  Name                 Priority
  ------------------------------
  live                   1-   9
  important             10-  19
  background            20-
  ------------------------------
--]]
-- funtion: enqueue_request
-- argument: map, x, y, zoom, priority
--     priority = 1-10 for live requests
-- return:   true or nil
--
function enqueue_request (map, x, y, z, priority)
    local mx = x - x % 8
    local my = y - y % 8
    local mz = z
    local id = time()
    local priority = tonumber(priority)
    local index = get_key(map, mx, my, mz)
    local req = serialize_msg({
        ["id"]   = tostring(id);
        ["type"] = 'metatile_enqueue_request';
        ["prio"] = priority;
        ["map"]  = map;
        ["x"]    = mx;
        ["y"]    = my;
        ["z"]    = mz})
    local ok = get_handle(index, req, 300, GOTHANDLE)
    if not ok then
        return wait_signal(index, 30)
    end
    local msg = send_tirex_request(req)
    if not msg then
        return send_signal(index, 300, FAILED)
    end
    local index = get_key(msg["map"], msg["x"], msg["y"], msg["z"])
    local res = msg["result"]
    if res == "ok" then
        return send_signal(index, 300, SUCCEEDED)
    else
        return send_signal(index, 300, FAILED)
    end
end

-- funtion: dequeue_request
-- argument: map, x, y, z, priority
-- return:   true or nil
--
function dequeue_request (map, x, y, z, priority)
    local mx = x - x % 8
    local my = y - y % 8
    local mz = z
    local id = time()
    local priority = tonumber(priority)
    local index = get_key(map, mx, my, mz)
    local req = serialize_msg({
        ["id"]   = tostring(id);
        ["type"] = 'metatile_remove_request';
        ["prio"] = priority;
        ["map"]  = map;
        ["x"]    = mx;
        ["y"]    = my;
        ["z"]    = mz})
    local ok = get_handle(index, req, 300, GOTHANDLE)
    if not ok then
        return wait_signal(index, 30)
    end
    local msg = send_tirex_request(req)
    if msg then
        local res = msg["result"]
        if res == "ok" then
            return send_signal(index, 300, SUCCEEDED)
        else
            return send_signal(index, 300, FAILED)
        end
    end
end

-- function: ping()
-- return: true or nil
function ping()
    -- Create request command
    local req = serialize_msg({["type"] = 'ping'})
    local msg = send_tirex_request(req)
    if not msg then
        return nil
    end
    if msg["result"] ~= 'ok' then
        return nil
    end
    return true
end

-- funtion: enqueue_request_with_larger_zoom
-- argument: map, x, y, zoom, maxzoom, priority
-- return:   true or nil
--
function request (map, x, y, z1, z2, priority)
    local z2 = tonumber(z2)
    local z1 = tonumber(z1)
    if z1 > z2 then
        return nil
    end
    local priority = tonumber(priority)
    local np = priority + 10
    -- assume that live priority is processed on 'live' and is in 1-9 range.
    -- and larger zoom rendering is on 'important' in 10-19 range.
    local res = enqueue_request(map, x, y, z1, priority)
    if not res then
        return nil
    end
    if z1 == z2 then
        return true
    end
    for i = 1, z2 - z1 do
        local nx, ny = osm_tile.zoom_num(x, y, z1, z1 + i)
        background_enqueue_request(map, nx, ny, z1 + i, np + i)
    end
    return true
end

function cancel (map, x, y, z1, z2, prirority)
    local z2 = tonumber(z2)
    local z1 = tonumber(z1)
    if z1 > z2 then
        return nil
    end
    local priority = tonumber(priority)
    for i = 0, z2 - z1 do
        local nx, ny = osm_tile.zoom_num(x, y, z1, z1 + i)
        dequeue_request(map, nx, ny, z1 + i, priority)
    end
    return true
end

local class_mt = {
    -- to prevent use of casual module global variables
    __newindex = function (table, key, val)
        error('attempt to write to undeclared variable "' .. key .. '"')
    end
}

setmetatable(_M, class_mt)
